File: dohproxy/https_proxy.py
"""Transparent TLS-splitting relay for outbound TCP/443 (SNI/DPI bypass).

Why a relay and not raw packet surgery: defeating an SNI filter that reassembles
the TCP stream requires TLS *record-layer* fragmentation, which inserts a second
5-byte record header into the byte stream. You cannot insert bytes on the raw
WinDivert path without desyncing the client kernel's TCP sequence numbers (the
kernel would RST when the server ACKs bytes it never sent). So, exactly like
Intra, we terminate the connection and reframe it from a process that owns both
sockets.

Wiring reuses the WinDivert redirect trick from `tcp_proxy.py`:
  * Outbound client->server :443 packet -> remember the real server keyed by
    (src_addr, src_port), rewrite the destination to src_addr:HTTPS_PROXY_PORT,
    and inject it INBOUND so the local stack hands it to our listener.
  * relay->client packet (src port == HTTPS_PROXY_PORT) -> rewrite the source
    back to server:443 so the client's socket accepts it, inject INBOUND.

The relay's upstream sockets (relay -> real server) are bound to a reserved
source-port range that the kernel filter excludes, so they travel normally and
are never re-captured (no inject loop, no added capture cost on that leg).

Packet rewriting runs inline on the capture thread (so `_conn_map` needs no
lock); each accepted connection is then served on its own thread.
"""

from __future__ import annotations

import logging
import socket
import socketserver
import threading

from pydivert.consts import Direction

from . import config, dpi

log = logging.getLogger("dohproxy.https")

# (src_addr, src_port) -> (orig_dst_addr, orig_dst_port).
# Written by the capture thread (every client->server packet refreshes it) and
# read by handler threads. Refresh-on-every-packet makes stale entries harmless.
_conn_map: dict[tuple[str, int], tuple[str, int]] = {}


# --------------------------------------------------------------------------- #
# Packet rewriting (capture thread)
# --------------------------------------------------------------------------- #
def handle_packet(packet, send) -> None:
    if packet.is_outbound and packet.dst_port == 443:
        # The relay's own upstream sockets are excluded by the kernel filter, so
        # anything reaching here is a real client connection to redirect.
        _redirect(packet, send)
    elif packet.src_port == config.HTTPS_PROXY_PORT:
        _rewrite_reply(packet, send)
    else:
        send(packet)


def _redirect(packet, send) -> None:
    key = (packet.src_addr, packet.src_port)
    _conn_map[key] = (packet.dst_addr, packet.dst_port)
    if packet.tcp.rst:
        _conn_map.pop(key, None)

    packet.dst_addr = packet.src_addr
    packet.dst_port = config.HTTPS_PROXY_PORT
    packet.direction = Direction.INBOUND
    send(packet)


def _rewrite_reply(packet, send) -> None:
    key = (packet.dst_addr, packet.dst_port)
    server = _conn_map.get(key)
    if server is None:
        return  # unknown/teardown stray; drop
    packet.src_addr, packet.src_port = server
    packet.direction = Direction.INBOUND
    send(packet)

    if packet.tcp.rst or packet.tcp.fin:
        _conn_map.pop(key, None)


# --------------------------------------------------------------------------- #
# Reserved upstream source ports (so the relay's upstream leg is never captured)
# --------------------------------------------------------------------------- #
_port_lock = threading.Lock()
_next_port = config.UPSTREAM_PORT_BASE


def _connect_upstream(server_ip: str, server_port: int) -> socket.socket:
    """Open an upstream socket bound to a port in the reserved range and
    connected to the real server."""
    global _next_port
    base = config.UPSTREAM_PORT_BASE
    hi = base + config.UPSTREAM_PORT_COUNT
    last_err: Exception | None = None

    for _ in range(config.UPSTREAM_PORT_COUNT):
        with _port_lock:
            port = _next_port
            _next_port = port + 1 if port + 1 < hi else base

        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        try:
            s.bind(("0.0.0.0", port))
        except OSError as exc:  # port busy -> try the next one
            last_err = exc
            s.close()
            continue
        try:
            s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
            s.settimeout(config.HTTPS_CONNECT_TIMEOUT)
            s.connect((server_ip, server_port))
            s.settimeout(None)
            return s
        except OSError as exc:
            last_err = exc
            s.close()
            raise

    raise OSError(f"no free upstream port in reserved range ({last_err})")


# --------------------------------------------------------------------------- #
# Local relay server
# --------------------------------------------------------------------------- #
def _pump(src: socket.socket, dst: socket.socket) -> None:
    """Copy src -> dst until EOF, then half-close dst's write side."""
    try:
        while True:
            data = src.recv(65535)
            if not data:
                break
            dst.sendall(data)
    except OSError:
        pass
    finally:
        try:
            dst.shutdown(socket.SHUT_WR)
        except OSError:
            pass


class _Handler(socketserver.BaseRequestHandler):
    def handle(self) -> None:
        client = self.request
        peer = self.client_address
        # Redirected connections always have peer IP == this host's IP. Reject
        # anything else so we never act as an open proxy.
        if peer[0] != client.getsockname()[0]:
            return

        orig = _conn_map.get((peer[0], peer[1]))
        if orig is None:
            return
        server_ip, server_port = orig

        try:
            client.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        except OSError:
            pass

        try:
            upstream = _connect_upstream(server_ip, server_port)
        except OSError as exc:
            log.warning("[HTTPS] upstream %s:%d failed: %s", server_ip, server_port, exc)
            return

        try:
            self._relay(client, upstream, server_ip, server_port)
        finally:
            upstream.close()

    def _relay(self, client, upstream, server_ip, server_port) -> None:
        # Read the first client segment -- the TLS ClientHello -- and re-emit it
        # fragmented across two TLS records.
        client.settimeout(config.HTTPS_FIRST_READ_TIMEOUT)
        try:
            first = client.recv(65535)
        except (socket.timeout, OSError):
            return
        client.settimeout(None)
        if not first:
            return

        try:
            if first[0] == dpi._TLS_HANDSHAKE:
                segs = dpi.split_hello(first, config.SPLIT_MIN, config.SPLIT_MAX)
                log.info(
                    "[HTTPS] %s:%d  SNI=%s  ClientHello %dB -> %d TLS records",
                    server_ip, server_port, dpi.sni_name(first), len(first), len(segs),
                )
                for seg in segs:
                    upstream.sendall(seg)
            else:
                # Not TLS (e.g. plaintext on 443): forward untouched.
                upstream.sendall(first)
        except OSError as exc:
            log.debug("[HTTPS] %s:%d first write failed: %s", server_ip, server_port, exc)
            return

        # Dumb bidirectional pipe for the rest of the connection.
        reverse = threading.Thread(
            target=_pump, args=(upstream, client), name="https-pump", daemon=True
        )
        reverse.start()
        _pump(client, upstream)
        reverse.join(timeout=2.0)


class _Server(socketserver.ThreadingTCPServer):
    allow_reuse_address = True
    daemon_threads = True


def start_server() -> socketserver.ThreadingTCPServer:
    server = _Server((config.TCP_BIND_HOST, config.HTTPS_PROXY_PORT), _Handler)
    threading.Thread(target=server.serve_forever, name="https-proxy", daemon=True).start()
    log.info(
        "HTTPS splitting relay listening on %s:%d (upstream ports %d-%d)",
        config.TCP_BIND_HOST, config.HTTPS_PROXY_PORT,
        config.UPSTREAM_PORT_BASE, config.UPSTREAM_PORT_BASE + config.UPSTREAM_PORT_COUNT - 1,
    )
    return server

File: src/https_proxy.rs
//! Transparent TLS-splitting relay for outbound TCP/443 (SNI/DPI bypass).
//!
//! Defeating an SNI filter that reassembles the TCP stream requires TLS
//! *record-layer* fragmentation, which inserts a second 5-byte record header into
//! the byte stream. You cannot insert bytes on the raw WinDivert path without
//! desyncing the client kernel's TCP sequence numbers, so -- exactly like Intra --
//! we terminate each :443 connection at a tiny local relay (the same redirect
//! trick as tcp_proxy) and reframe the ClientHello from a process that owns both
//! sockets.
//!
//! The relay's upstream sockets bind to a reserved source-port range the kernel
//! filter excludes, so they travel normally and are never re-captured.
//!
//! Packet rewriting runs on the capture thread; the connection map is shared with
//! handler threads, so it is behind a Mutex (the Python relied on the GIL).

use std::collections::HashMap;
use std::io::{Read, Write};
use std::net::{Ipv4Addr, SocketAddr, TcpListener, TcpStream};
use std::sync::atomic::{AtomicU32, Ordering};
use std::sync::{Mutex, OnceLock};

use anyhow::{anyhow, Context, Result};
use socket2::{Domain, Protocol, Socket, Type};
use windivert::prelude::*;

use crate::config;
use crate::divert::{inject_inbound, Injector};
use crate::netpkt::{self, TCP_FIN, TCP_RST};
use crate::tcp_proxy::ConnMap;
use crate::dpi;

// (src_addr, src_port) -> (orig_dst_addr, orig_dst_port). Written by the capture
// thread, read by handler threads -> Mutex.
static CONN: OnceLock<Mutex<ConnMap>> = OnceLock::new();

fn conn() -> &'static Mutex<ConnMap> {
    CONN.get_or_init(|| Mutex::new(HashMap::new()))
}

// --------------------------------------------------------------------------- //
// Packet rewriting (capture thread)
// --------------------------------------------------------------------------- //
pub fn handle_packet(packet: &mut WinDivertPacket<NetworkLayer>, inj: &Injector) {
    let outbound = packet.address.outbound();
    let dst_port = netpkt::l4_dst_port(&packet.data);
    let src_port = netpkt::l4_src_port(&packet.data);

    if outbound && dst_port == Some(443) {
        redirect(packet, inj);
    } else if src_port == Some(config::HTTPS_PROXY_PORT) {
        rewrite_reply(packet, inj);
    } else {
        inj.send(packet);
    }
}

fn redirect(packet: &mut WinDivertPacket<NetworkLayer>, inj: &Injector) {
    let (Some(src), Some(sport)) = (netpkt::src_addr(&packet.data), netpkt::l4_src_port(&packet.data))
    else {
        return;
    };
    let (Some(dst), Some(dport)) = (netpkt::dst_addr(&packet.data), netpkt::l4_dst_port(&packet.data))
    else {
        return;
    };
    let key = (src, sport);
    {
        let mut map = conn().lock().unwrap();
        map.insert(key, (dst, dport));
        if netpkt::tcp_flags(&packet.data) & TCP_RST != 0 {
            map.remove(&key);
        }
    }

    let data = packet.data.to_mut();
    netpkt::set_dst_addr(data, src);
    netpkt::set_l4_dst_port(data, config::HTTPS_PROXY_PORT);
    inject_inbound(inj, packet);
}

fn rewrite_reply(packet: &mut WinDivertPacket<NetworkLayer>, inj: &Injector) {
    let (Some(dst), Some(dport)) = (netpkt::dst_addr(&packet.data), netpkt::l4_dst_port(&packet.data))
    else {
        return;
    };
    let key = (dst, dport);
    let server = conn().lock().unwrap().get(&key).copied();
    let Some((sip, sport)) = server else {
        return; // unknown/teardown stray; drop
    };

    let flags = netpkt::tcp_flags(&packet.data);
    let data = packet.data.to_mut();
    netpkt::set_src_addr(data, sip);
    netpkt::set_l4_src_port(data, sport);
    inject_inbound(inj, packet);

    if flags & (TCP_RST | TCP_FIN) != 0 {
        conn().lock().unwrap().remove(&key);
    }
}

// --------------------------------------------------------------------------- //
// Reserved upstream source ports (so the relay's upstream leg is never captured)
// --------------------------------------------------------------------------- //
static NEXT_PORT: AtomicU32 = AtomicU32::new(0);

fn connect_upstream(server_ip: Ipv4Addr, server_port: u16) -> Result<TcpStream> {
    let base = config::UPSTREAM_PORT_BASE;
    let count = config::UPSTREAM_PORT_COUNT as u32;
    let mut last_err: Option<std::io::Error> = None;

    for _ in 0..count {
        let n = NEXT_PORT.fetch_add(1, Ordering::Relaxed);
        let port = base + (n % count) as u16;

        let socket = Socket::new(Domain::IPV4, Type::STREAM, Some(Protocol::TCP))
            .context("creating upstream socket")?;
        let _ = socket.set_reuse_address(true);
        let bind_addr: SocketAddr = (Ipv4Addr::UNSPECIFIED, port).into();
        if let Err(e) = socket.bind(&bind_addr.into()) {
            last_err = Some(e); // port busy -> try the next one
            continue;
        }
        let _ = socket.set_nodelay(true);
        let server: SocketAddr = (server_ip, server_port).into();
        // Connect failures propagate immediately (mirrors the Python relay).
        socket
            .connect_timeout(&server.into(), config::HTTPS_CONNECT_TIMEOUT)
            .with_context(|| format!("connecting upstream {server_ip}:{server_port}"))?;
        return Ok(socket.into());
    }

    Err(anyhow!(
        "no free upstream port in reserved range ({last_err:?})"
    ))
}

// --------------------------------------------------------------------------- //
// Local relay server
// --------------------------------------------------------------------------- //
fn pump(mut src: TcpStream, mut dst: TcpStream) {
    let mut buf = vec![0u8; 65535];
    loop {
        match src.read(&mut buf) {
            Ok(0) | Err(_) => break,
            Ok(n) => {
                if dst.write_all(&buf[..n]).is_err() {
                    break;
                }
            }
        }
    }
    let _ = dst.shutdown(std::net::Shutdown::Write);
}

fn serve_client(client: TcpStream) {
    let (Ok(peer), Ok(local)) = (client.peer_addr(), client.local_addr()) else {
        return;
    };
    // Redirected connections always have peer IP == this host's IP. Reject
    // anything else so we never act as an open proxy.
    if peer.ip() != local.ip() {
        return;
    }

    let key = match peer {
        SocketAddr::V4(a) => (*a.ip(), a.port()),
        _ => return,
    };
    let Some((server_ip, server_port)) = conn().lock().unwrap().get(&key).copied() else {
        return;
    };

    let _ = client.set_nodelay(true);

    let upstream = match connect_upstream(server_ip, server_port) {
        Ok(u) => u,
        Err(e) => {
            log::warn!(target: "freegsm.https",
                "[HTTPS] upstream {server_ip}:{server_port} failed: {e:#}");
            return;
        }
    };

    relay(client, upstream, server_ip, server_port);
}

fn relay(mut client: TcpStream, mut upstream: TcpStream, server_ip: Ipv4Addr, server_port: u16) {
    // Read the first client segment -- the TLS ClientHello -- and re-emit it
    // fragmented across two TLS records.
    let _ = client.set_read_timeout(Some(config::HTTPS_FIRST_READ_TIMEOUT));
    let mut buf = vec![0u8; 65535];
    let first_len = match client.read(&mut buf) {
        Ok(0) | Err(_) => return,
        Ok(n) => n,
    };
    let _ = client.set_read_timeout(None);
    let first = &buf[..first_len];

    if first[0] == dpi::TLS_HANDSHAKE {
        let segs = dpi::split_hello(first, config::SPLIT_MIN, config::SPLIT_MAX);
        log::info!(target: "freegsm.https",
            "[HTTPS] {server_ip}:{server_port}  SNI={}  ClientHello {}B -> {} TLS records",
            dpi::sni_name(first), first.len(), segs.len());
        for seg in &segs {
            if upstream.write_all(seg).is_err() {
                return;
            }
        }
    } else {
        // Not TLS (e.g. plaintext on 443): forward untouched.
        if upstream.write_all(first).is_err() {
            return;
        }
    }

    // Dumb bidirectional pipe for the rest of the connection.
    let (Ok(client_rx), Ok(upstream_rx)) = (client.try_clone(), upstream.try_clone()) else {
        return;
    };
    let reverse = std::thread::Builder::new()
        .name("https-pump".into())
        .spawn(move || pump(upstream_rx, client_rx));
    pump(client, upstream);
    if let Ok(handle) = reverse {
        let _ = handle.join();
    }
}

/// Start the HTTPS splitting relay on its own thread.
pub fn start_server() -> Result<()> {
    let listener = TcpListener::bind((config::TCP_BIND_HOST, config::HTTPS_PROXY_PORT))
        .with_context(|| {
            format!(
                "binding HTTPS relay on {}:{}",
                config::TCP_BIND_HOST,
                config::HTTPS_PROXY_PORT
            )
        })?;
    log::info!(target: "freegsm.https",
        "HTTPS splitting relay listening on {}:{} (upstream ports {}-{})",
        config::TCP_BIND_HOST, config::HTTPS_PROXY_PORT,
        config::UPSTREAM_PORT_BASE,
        config::UPSTREAM_PORT_BASE + config::UPSTREAM_PORT_COUNT - 1);
    std::thread::Builder::new()
        .name("https-proxy".into())
        .spawn(move || {
            for stream in listener.incoming() {
                match stream {
                    Ok(s) => {
                        std::thread::spawn(move || serve_client(s));
                    }
                    Err(e) => log::debug!(target: "freegsm.https", "accept error: {e}"),
                }
            }
        })
        .context("spawning https-proxy thread")?;
    Ok(())
}
